File: 2805_CutTree.py
n,m=map(int,input().split())
tree=sorted(list(map(int,input().split())))

left,right=0,tree[n-1]
ans=0
while left<=right:
    mid=(left+right)//2
    cut=0
    for n in tree:
        if n>mid:
            cut+=n-mid
    
    if cut>=m:
        ans=max(ans,mid) # 나무의 길이가 최대일 때
        left=mid+1
    else:
        right=mid-1

print(ans)

File: 2805_CutTree.cpp
#include<iostream>
#include<algorithm>
#define MAX 1000000

int tree[MAX];
long long dp[MAX];
using namespace std;
int main() {
	int n, m;
	cin >> n >> m;

	for (int i = 0; i < n; i++)
		cin >> tree[i];

	sort(tree, tree + n);

	int cut = 0;
	dp[n - 1] = 0;
	for (int i = n - 2; i >= 0; i--) {
		dp[i] = dp[i + 1] + (tree[i + 1] - tree[i]) * (n - i - 1);
		if (dp[i] > m) {
			cut = i + 1;
			break;
		}
		else if (dp[i] == m) {
			cut = i;
			break;
		}
	}

	int ans = tree[cut];
	long long remain = dp[cut];
	while (remain < m) {
		remain += 1 * (n - cut);
		ans--;
	}
	
	cout << ans << "\n";

	return 0;
}
